<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логическая-регрессия" data-toc-modified-id="Логическая-регрессия-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Логическая регрессия</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Тестирование-&quot;Дерево-решений&quot;" data-toc-modified-id="Тестирование-&quot;Дерево-решений&quot;-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Тестирование "Дерево решений"</a></span></li><li><span><a href="#Тестирование-&quot;Случайный-лес&quot;" data-toc-modified-id="Тестирование-&quot;Случайный-лес&quot;-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Тестирование "Случайный лес"</a></span></li><li><span><a href="#Тестирование-&quot;Логическая-регрессия&quot;" data-toc-modified-id="Тестирование-&quot;Логическая-регрессия&quot;-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Тестирование "Логическая регрессия"</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [66]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

In [67]:
data = pd.read_csv('/datasets/users_behavior.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [68]:
data.head(10)

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0

В таблице информация по 3214 пользователям с количеством звонков, минут, мегабайт и столбец, указывающий является ли пользователь владельцем тарифа Ультра. 1-Ультра, 0-Смарт.

## Разбейте данные на выборки

In [69]:
data_train, data_other = train_test_split(data, test_size=0.25, random_state=12345)
data_test, data_valid = train_test_split(data_other, test_size=0.5, random_state=12345)

print(f'Размер тестовой выборки: {data_test.shape}')
print(f'Размер тренировочной выборки: {data_train.shape}')
print(f'Размер валидационной выборки: {data_valid.shape}')

Размер тестовой выборки: (402, 5)
Размер тренировочной выборки: (2410, 5)
Размер валидационной выборки: (402, 5)


In [70]:
#
#train
features_train = data_train.drop('is_ultra', axis=1)
target_train = data_train['is_ultra']

#test
features_test = data_test.drop('is_ultra', axis=1)
target_test = data_test['is_ultra']

#valid
features_valid = data_valid.drop('is_ultra', axis=1)
target_valid = data_valid['is_ultra']

## Исследуйте модели

### Дерево решений

In [71]:
for depth in range(1,26):
    model_tree = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model_tree.fit(features_train, target_train)
    prediction_tree = model_tree.predict(features_valid)
    accuracy_tree = accuracy_score(prediction_tree, target_valid)
    print('Max_depth:', depth, end='')
    print(' Accuracy:', accuracy_tree)

Max_depth: 1 Accuracy: 0.736318407960199
Max_depth: 2 Accuracy: 0.7736318407960199
Max_depth: 3 Accuracy: 0.7786069651741293
Max_depth: 4 Accuracy: 0.763681592039801
Max_depth: 5 Accuracy: 0.763681592039801
Max_depth: 6 Accuracy: 0.753731343283582
Max_depth: 7 Accuracy: 0.7910447761194029
Max_depth: 8 Accuracy: 0.7935323383084577
Max_depth: 9 Accuracy: 0.7985074626865671
Max_depth: 10 Accuracy: 0.7935323383084577
Max_depth: 11 Accuracy: 0.7960199004975125
Max_depth: 12 Accuracy: 0.7835820895522388
Max_depth: 13 Accuracy: 0.7786069651741293
Max_depth: 14 Accuracy: 0.763681592039801
Max_depth: 15 Accuracy: 0.7711442786069652
Max_depth: 16 Accuracy: 0.7587064676616916
Max_depth: 17 Accuracy: 0.7437810945273632
Max_depth: 18 Accuracy: 0.7313432835820896
Max_depth: 19 Accuracy: 0.7189054726368159
Max_depth: 20 Accuracy: 0.7388059701492538
Max_depth: 21 Accuracy: 0.7263681592039801
Max_depth: 22 Accuracy: 0.7164179104477612
Max_depth: 23 Accuracy: 0.7189054726368159
Max_depth: 24 Accuracy: 0

При Max_depth = 9, качество модели "Дерево решений" ~ 0.8  


### Случайный лес

In [72]:
for est in range(1,26):
    model_forest = RandomForestClassifier(n_estimators=est, random_state=12345, max_depth=10)
    model_forest.fit(features_train, target_train)
    prediction_forest = model_forest.predict(features_valid)
    accuracy_forest = accuracy_score(prediction_forest, target_valid)
    print('n_estimators:', est, end='')
    print(' Accuracy:', accuracy_forest)

n_estimators: 1 Accuracy: 0.7761194029850746
n_estimators: 2 Accuracy: 0.763681592039801
n_estimators: 3 Accuracy: 0.7860696517412935
n_estimators: 4 Accuracy: 0.7885572139303483
n_estimators: 5 Accuracy: 0.7885572139303483
n_estimators: 6 Accuracy: 0.8009950248756219
n_estimators: 7 Accuracy: 0.7885572139303483
n_estimators: 8 Accuracy: 0.7960199004975125
n_estimators: 9 Accuracy: 0.7985074626865671
n_estimators: 10 Accuracy: 0.7985074626865671
n_estimators: 11 Accuracy: 0.8034825870646766
n_estimators: 12 Accuracy: 0.8059701492537313
n_estimators: 13 Accuracy: 0.7985074626865671
n_estimators: 14 Accuracy: 0.8084577114427861
n_estimators: 15 Accuracy: 0.8134328358208955
n_estimators: 16 Accuracy: 0.8109452736318408
n_estimators: 17 Accuracy: 0.8084577114427861
n_estimators: 18 Accuracy: 0.8084577114427861
n_estimators: 19 Accuracy: 0.8109452736318408
n_estimators: 20 Accuracy: 0.8034825870646766
n_estimators: 21 Accuracy: 0.8059701492537313
n_estimators: 22 Accuracy: 0.803482587064676

При n_estimators = 15, качество модели "Случайный лес" ~ 0.81

### Логическая регрессия

In [73]:
model_log_regression = LogisticRegression()
model_log_regression.fit(features_train, target_train)
predict_log_regression = model_log_regression.predict(features_valid)
accuracy_log_regression = accuracy_score(predict_log_regression, target_valid)
print(f'Accuracy: {accuracy_log_regression}')

Accuracy: 0.7039800995024875


Обучив три разные модели, получили лучшие показтели качества:   
• Дерево решений, при гиперпараметре max_depth = 9 → 0.8    
• Случайный лес, при гиперпараметре max_depth = 10, n_estimators = 15 → 0.81    
• Логистическая регрессия → 0.7    
▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬   
Лучшая модель → Случайный лес


## Проверьте модель на тестовой выборке

### Тестирование "Дерево решений"

In [74]:
model_tree = DecisionTreeClassifier(random_state=12345, max_depth=9)
model_tree.fit(features_train, target_train)
prediction_tree = model_tree.predict(features_test)
accuracy_tree = accuracy_score(prediction_tree, target_test)
print(f'Accuracy: {accuracy_tree}')

Accuracy: 0.7786069651741293


### Тестирование "Случайный лес"

In [75]:
model_forest = RandomForestClassifier(random_state=12345, max_depth=10, n_estimators=15)
model_forest.fit(features_train, target_train)
prediction_forest = model_forest.predict(features_test)
accuracy_forest = accuracy_score(prediction_forest, target_test)
print(f'Accuracy: {accuracy_forest}')

Accuracy: 0.8034825870646766


### Тестирование "Логическая регрессия"

In [76]:
model_log_regression = LogisticRegression()
model_log_regression.fit(features_train, target_train)
predict_log_regression = model_log_regression.predict(features_test)
accuracy_log_regression = accuracy_score(predict_log_regression, target_test)
print(f'Accuracy: {accuracy_log_regression}')

Accuracy: 0.7039800995024875


Результаты тестирования соответствуют предыдущим результатам. Лучшей моделью по метрике Accuracy остается модель "Случайный лес". 

## (бонус) Проверьте модели на адекватность

In [77]:
random_results = np.random.randint(0, 2, size=402)

In [78]:
random_test = accuracy_score(target_test, random_results)
random_test

0.48507462686567165

Модель "Случайный лес" дает намного больше верных ответов, чем случайные предсказания.

## Вывод

Мы рассмотрели три модели "Случайный лес", "Логическая регрессия", "Дерево решений". Меняя их гиперпараметры мы нашли вариант с наибольшим количеством верных ответов. Лучше всего подходит модель "Случайный лес" с показателем Accuracy примерно 0.81.   
Полученная модель дает почти в два раза больше верных ответов, чем случайные предсказания. 